mtsp/src/main.leo

```diff
@@ -0,0 +1,673 @@
+  record Token {
+    owner: address,
+    amount: u128,
+    token_id: field,
+    authorized: bool,
+    authorized_until: u32
+  }
+
+  struct TokenMetadata {
+    token_id: field,
+    name: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    symbol: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    decimals: u8,
+    supply: u128,
+    max_supply: u128,
+    admin: address,
+    external_authorization_required: bool // whether or not this token requires authorization from an external program before transferring
+  }
+
+  struct TokenOwner {
+    account: address,
+    token_id: field
+  }
+
+  struct Balance {
+    token_id: field,
+    account: address,
+    balance: u128,
+    authorized_until: u32
+  }
+
+  struct Allowance {
+    account: address,
+    spender: address,
+    token_id: field
+  }
+
+  mapping registered_tokens: field => TokenMetadata;
+  mapping balances: field => Balance; // hash(token_id, account) => Balance
+  mapping authorized_balances: field => Balance; // hash(token_id, account) => Balance
+  mapping allowances: field => u128; // hash(token_id, account, spender) => Allowance
+
+  // -------------------------
+  // Called by token admins
+  // -------------------------
+
+  transition register_token(token_id: field, name: u128, symbol: u128, decimals: u8, max_supply: u128, external_authorization_required: bool) {
+    let token: TokenMetadata = TokenMetadata {
+      token_id: token_id,
+      name: name,
+      symbol: symbol,
+      decimals: decimals,
+      supply: 0u128,
+      max_supply: max_supply,
+      admin: self.caller,
+      external_authorization_required: external_authorization_required
+    };
+    return then finalize(token);
+  }
+
+  finalize register_token(token: TokenMetadata) {
+    // Make sure token doesn't already exist
+    let token_exists: bool = registered_tokens.contains(token.token_id);
+    assert_eq(token_exists, false);
+    // Set new token
+    registered_tokens.set(token.token_id, token);
+  }
+
+  // TODO: update_token_metadata
+
+  transition mint_public(token_id: field, recipient: address, amount: u128, authorized_until: u32) {
+    return then finalize(token_id, recipient, amount, authorized_until, self.caller);
+  }
+
+  finalize mint_public(token_id: field, recipient: address, amount: u128, authorized_until: u32, token_admin: address) {
+    // Check that the token exists, and that the caller is the token admin
+    // Check that the token supply + amount <= max_supply
+    let token: TokenMetadata = registered_tokens.get(token_id);
+    assert_eq(token.admin, token_admin);
+    let new_supply: u128 = token.supply + amount;
+    assert(new_supply <= token.max_supply);
+
+    // Get or create the balance for the recipient
+    let token_owner: TokenOwner = TokenOwner {
+      account: recipient,
+      token_id: token_id
+    };
+    let balance_key: field = BHP256::hash_to_field(token_owner);
```

Comment:
Is there a reason why we do this pattern of hashing for the mapping key, instead of directly using the struct?